Added data triggers, refactored, and removed ExtendedAbilityBehaviour

divisionbyz0rro opened the pull request:
This PR does three main things:

1. The 'TriggerIdentification' class has been removed in favor of using GuidManager to extend CustomTrigger whenever a modder wants to implement a custom trigger. This significantly cuts down on the amount of code required.
2. The definition of the method signature for each trigger is now managed as a MethodInfo instead of a List<Type>. This MethodInfo is derived from a delegate, which is kept in an extra file. This both supports item (3) on this list, and also gives us a clear place to put XML documentation to clearly document how each of the custom triggers are supposed to behave.
3. Adds support for custom triggers that don't return IEnumerator and instead return an arbitrary type. This is to provide a generic hook for the global trigger handler to ask all cards to calculate some sort of metadata and return it. There are three of these custom data triggers included in this PR: OnGetOpposingSlots (calculates which slots the card should attack), and the pair of OnBuffOtherCardAttack/OnBuffOtherCardHealth (which support custom passive attack/health buffs).

Along with these, ExtendedAbilityBehaviour is now officially retired.

Note that the patches that make OnBuffOtherCardAttack/Health work are actually not activated right now. This is due to their horrific performance when they run every frame. They should be uncommented once we fix that problem.

SpecialAPI commented:
The support for triggers with return types other than IEnumerator is a very good feature, however I don't really like extending custom enums with GuidManager, that is the whole reason I added trigger identification, so imo it should stay

divisionbyz0rro commented:
> The support for triggers with return types other than IEnumerator is a very good feature, however I don't really like extending custom enums with GuidManager, that is the whole reason I added trigger identification, so imo it should stay

GuidManager was of course originally written to extend enums that only DM has control over as opposed to ones that we do, but from the perspective of the modder using this API there's no difference. Whether you want to add a new Ability or a new CustomTrigger, the guidance is always the same: use GuidManager. It's crude, but it serves its purpose and keeps the codebase as tight as possible.

FWIW: I almost just suggested dropping CustomTrigger altogether and managing everything as an extension of the base game's Trigger enum. The only reason I didn't do that was because that could lead to confusion; sending one of the extended triggers into the base game's trigger manger could potentially create issues that I don't think we want to write patches to fix.

That being said: I did have to update the attributes just now. I realized that I made a mistake - only constants can go into the parameters of an attribute, so it's not actually possible to use the result of a call to GuidManager in an attribute contructor. So I added back constructors to each attribute that accept a mod guid and trigger name but still call GuidManager behind the scenes.

divisionbyz0rro commented:
> but why? unlike adding stuff to basegame, where we have no control over how it's identified and we have to use guidmanager, this is code that we CAN modify, so i think just keeping it as two strings is better than doing enum extensions with guidmanager. there are really no reasons i see (except making api have less code) to actually use guidmanager instead of the current solution.

**We** can modify CustomTrigger, but the modders can't. From their perspective its still an Enum that they have no control over. Enums you can't control get managed by GuidManager.

I just don't see a compelling reason *not* to use the tool we already have at our disposal. It adds consistency and reduces the overall size of the codebase. And as far as I can see, there's no downside to using it.

divisionbyz0rro commented:
> > I just don't see a compelling reason _not_ to use the tool we already have at our disposal. It adds consistency and reduces the overall size of the codebase. And as far as I can see, there's no downside to using it.
> i don't understand your point at all. changing it to guid manager doesn't do anything on moddev side it feels wrong to me. if you're worrying about documentation you won't need to document TriggerIdentification anyways. for the duplicate methods that require strings we can just copy-paste the descriptions of existing things and change them. for readability it doesn't seem so bad. and i completely don't see any point in sacrificing a (in my opinion) better solution to a problem to reduce the code size a little (i know it's not just a little but still).

I guess I don't really understand what "feels" wrong about it to you.

The way I see it: we have a tool that we use when we want to extend an enumeration. When the modder wants to add a new CustomTrigger, we should use that tool. It simplifies the code and keeps things consistent.

If there's a specific downside to this that I'm missing, I'm more than happy to reconsider. But I see no downside - only upside.

@Windows10CE @julian-perge Thoughts on this discussion?

divisionbyz0rro commented:
It may be a moot point - Aaron wants to even further refactor the custom trigger handler stuff and probably wants to slap me for approving the original version in the first place. :)

divisionbyz0rro commented:
Closing this code review. I will open a new one once Aaron's rewrite is done, leveraging the Interface method.
